### DISCUSSION TITLE
Data Model Representations

### DISCUSSION THREAD

**Discussion by esoterra** [bytecodealliance/SIG-Registries#40] [General]
The registry's data model will need two (potentially same or different) representations.
1. The way data is abstractly encoded for the purpose of computing hashes
2. The way data is actually transported between clients and servers

It does not need to specify how data is internally stored by clients and servers; this is an implementation decision.

Some candidates for the abstract binary representation include
- CBOR
- JOSE
- Some custom format

Some candidates for the transport representation include
- CBOR
- JSON
- Protobuf

While we can choose different representations for each, doing so would require translating between them every time you send/receive data and then need to operate on it's defined abstract representation.

**Comment by lann** [comment]
> Some custom format

One approach I have used in personal projects is to use some existing serialization format but spec your message formats as if they were custom. I have done this with [BARE messages](https://baremessages.org/), where the encoding is simple enough to just write out for a few messages, followed by "this happens to exactly match this BARE schema: ...".

**Comment by esoterra** [comment]
I am currently using protobuf in the implementation as discussed with various members of the SIG. It has now been included in the updated proposal. If you have an objection to its use, please comment on the [proposal](https://github.com/bytecodealliance/SIG-Registries/pull/25).
